find.cc

```diff
@@ -526,12 +526,16 @@ class FindCommandParser {
     while (true) {
       if (!GetNextToken(&tok))
         return NULL;
-      if (tok != "-and" && tok != "-a") {
-        UngetToken(tok);
-        return c.release();
+      if (tok == "-and" || tok == "-a") {
+        if (!GetNextToken(&tok) || tok.empty())
+          return NULL;
+      } else {
+        if (tok != "-not" && tok != "\\!" && tok != "\\(" && tok != "-name" &&
```

Comment:
Checking if the next token is a prefix of `<fact>` would work now but in future it can be broken when the definition of `<fact>` has changed. Maybe better to do something like the following? I'm not very sure if my code works, but I think we shouldn't need two checks for `"-not"`, `"\\!"`, etc.

```
diff --git a/find.cc b/find.cc
index 1b9e23f..18450d0 100644
--- a/find.cc
+++ b/find.cc
@@ -526,15 +526,20 @@ class FindCommandParser {
     while (true) {
       if (!GetNextToken(&tok))
         return NULL;
-      if (tok != "-and" && tok != "-a") {
-        UngetToken(tok);
-        return c.release();
+      bool has_and = false;
+      if (tok == "-and" || tok == "-a") {
+        has_and = true;
+        if (!GetNextToken(&tok) || tok.empty())
+          return NULL;
       }
-      if (!GetNextToken(&tok) || tok.empty())
-        return NULL;
       unique_ptr<FindCond> r(ParseFact(tok));
       if (!r.get()) {
-        return NULL;
+        if (has_and) {
+          return NULL;
+        } else {
+          UngetToken(tok);
+          return c.release();
+        }
       }
       c.reset(new AndCond(c.release(), r.release()));
     }
```

Reply:
I had tried something similar, but ParseFact doesn't differentiate between an error and the token not being a fact. And that was necessary in order to parse things properly.